src/D2L.CodeStyle.Analyzers/Async/Generator/FileCollector.cs

```diff
@@ -0,0 +1,304 @@
+	/// namespaces, using directives, and type declarations for them to live in
+	/// (but only the necessary ones).
+	/// </summary>
+	internal sealed class FileCollector {
+		private static readonly SyntaxToken NoneToken
+			= SyntaxFactory.Token( SyntaxKind.None );
+
+		private static readonly SyntaxTriviaList EmptyTrivia
+			= SyntaxFactory.TriviaList();
+
+		private readonly CompilationUnitSyntax m_root;
+
+		// This is a mutable dictionary because we will remove from it as we go
+		// and check that it is empty at the end, just to be safe.
+		private readonly Dictionary<SyntaxNode, IEnumerable<string>> m_methods;
+
+		/// <summary>
+		/// Wraps the sort of syntax that we will append as part of a preamble, to
+		/// defer calls to ToFullString()
+		/// </summary>
+		private readonly struct PieceOfSyntax {
+			public object Data { get; init; }
+
+			public static implicit operator PieceOfSyntax( SyntaxNode node )
+				=> new() { Data = node };
+
+			public static implicit operator PieceOfSyntax( SyntaxToken token )
+				=> new() { Data = token };
+
+			public static implicit operator PieceOfSyntax( SyntaxList<UsingDirectiveSyntax> list )
+				=> new() { Data = list };
+
+			public override string ToString() => Data switch {
+				SyntaxNode n => n.ToFullString(),
+				SyntaxToken t => t.ToFullString(),
+				SyntaxList<UsingDirectiveSyntax> nl => nl.ToFullString(),
+				_ => throw new BugException( "invalid piece of syntax " + Data.GetType().Name )
+			};
+		}
+
+		// When we recurse into a namespace or type, we're not sure if the
+		// preamble ("namespace foo {", "\t\tpartial class Bar<T> {", etc.) is
+		// necessary, because maybe there are no generated methods.
+		// Using a LinkedList<string> as a double-ended queue.
+		private readonly LinkedList<PieceOfSyntax[]> m_preambles = new();
+
+		private readonly StringBuilder m_out = new();
+
+		private FileCollector(
+			CompilationUnitSyntax root,
+			Dictionary<SyntaxNode, IEnumerable<string>> methods
+		) {
+			m_root = root;
+			m_methods = methods;
+		}
+
+		public static FileCollector Create(
+			CompilationUnitSyntax root,
+			ImmutableArray<(SyntaxNode Parent, string Source)> methods
+		) {
+			var groupedMethods = methods
+				.GroupBy( static m => m.Parent )
+				.ToDictionary(
+					static g => g.Key,
+					static g => g.Select( static m => m.Source )
+				);
+
+			return new FileCollector( root, groupedMethods );
+		}
+
+		public string CollectSource() {
+			// File-scoped usings:
+			m_out.Append( m_root.Usings.ToFullString() );
+
+			WriteChildren( m_root );
+
+			if( m_preambles.Count != 0 ) {
+				throw new BugException( "left over preambles" );
+			}
+
+			if( m_methods.Count != 0 ) {
+				throw new BugException( "left over methods" );
+			}
+
+			return m_out.ToString();
+		}
+
+		private bool WriteChildren( SyntaxNode node ) {
+			// Early-out if we have no methods left to place.
+			if( m_methods.Count == 0 ) {
+				return false;
+			}

```

Comment:
Would this early out ever by appropriate first thing in `CollectSource` as well, or should FileCollector never be instantiated for nothing? Could that be a new BugException in create?

Reply:
I don't believe we'll ever actually get called with 0 methods, but I added a test to assert we output the empty string for that case (we might also output usings.)

This check would cause us to not consider Baz/Quux (not bother with their preambles) in this example:

```cs
partial class Foo {
  partial class Bar {
    [GenerateSync]
    async Task HelloAsync() => await Task.Delay( 3 );

    partial class Baz {}
  }

  partial class Quux {}
}
```

Reply:
> but I added a test to assert we output the empty string for that case (we might also output usings.)

Yeah, I was mostly thinking about that we'd be outputting a file of just usings, which isn't very helpful. But then we shouldn't be doing that anyway.

Reply:
Yeah. Like e2e I think that won't happen but I'm neutral on blocking it in here (imo intentionally crash or just do something "reasonable" like output an empty file/with usings)